affinity/cpp-23/dXXXXr0.md

```diff
@@ -0,0 +1,295 @@
+}  // std
+```
+*Listing 2: Addition to execution header synopsis*
+
+### Execution resources
+
+An *execution resource* represents an abstraction of a hardware or software
+layer that guarantees a particular set of affinity properties, where the level
+of abstraction is implementation-defined. An implementation is permitted to
+migrate any underlying resources providing it guarantees the affinity properties
+remain consistent. This allows freedom for the implementor but also consistency
+for users.
+
+If an *execution resource* is valid, then it must always point to the same
+underlying thing. For example, a *resource* cannot first point to one CPU core,
+and then suddenly point to a different CPU core. An *execution context* can thus
+rely on properties like binding of operating system threads to CPU cores.
+However, the "thing" to which an *execution resource* points may be a dynamic,
+possibly a software-managed pool of hardware. Here are three examples of this
+phenomenon:
+
+   1. The "hardware" may actually be a virtual machine (VM). At any point, the
+   VM may pause, migrate to different physical hardware, and resume. If the VM
+   presents the same virtual hardware before and after the migration, then the
+   *resources* that an application running on the VM sees should not change.
+   2. The OS may maintain a pool of a varying number of CPU cores as a shared
+   resource among different user-level processes. When a process stops using the
+   resource, the OS may reclaim cores. It may make sense to present this pool as
+   an *execution resource*.
+   3. A low-level device driver on a laptop may switch between a "discrete" GPU
+   and an "integrated" GPU, depending on utilization and power constraints. If
+   the two GPUs have the same instruction set and can access the same memory, it
+   may make sense to present them as a "virtualized" single *execution
+   resource*.
+
+In summary, an *execution resource* either identifies a thing uniquely, or
+harmlessly points to nothing.
+
+### Property group `adjacency`
+
+The `adjacency_t` properties are a group of mutually exclusive behavioral
+properties which provide a hint to an executor, via an *execution policy* which
+describes the implication of adjacent work-items of the callable of a bulk
+algorithm, allowing the implementation to choose the appropriate action.
+
+*Locality distance* specifies an implementation-defined metric for
+measuring the relative affinity between *execution resources* whereby *execution
+resources* with a lower *locality distance* are likely to have similar latency
+in memory access operations, for a given memory location.
+
+> [*Note:* For example, two hardware threads on the same CPU core would have a
+lower *locality distance* whereas two hardware threads on different CPU cores or
+in different NUMA regions would have a higher *locality distance*. *--end note*]
+
+> [*Note:* An alternative term of art for *locality distance* could be *locality
```

Comment:
"Distance" suggests comparability, e.g., in terms of minimum distance in a graph.  I think we want that ultimately, especially for introducing the various OpenMP arrangements that refer to different levels of a cache or node hierarchy, so it's not bad to save "distance" for later.

Reply:
That's a good point, do you think we should opt for "locality interference" as the term here?

Reply:
@AerialMantis I think it would be better to use "locality interference."  "Distance" suggests a metric with associated `operator<`, but I don't think we're intending to expose that here.  Even if we intend to expose affinity distances in the other proposal (where you can walk the hardware graph), affinity distances are a different thing than interference.  The former is a hardware property, while the latter is an algorithm property ("does the algorithm benefit from spatial locality?").

Would we ever intend to define a mapping between affinity distance and its benefits to constructive interference?

Reply:
I agree that locality distance is more of a property of the relationship between components of the topology, so perhaps makes more sense in the topology discovery interface, while locality interference is a property of the algorithm.

Yeah that's a good point, I think as we come to define a vocabulary for describing a system topology it makes sense to relate these to the algorithm properties, so we should keep this in mind.